contrib/graph-gophers/graphql-go/graphql.go

```diff
@@ -0,0 +1,72 @@
+package graphql
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/graph-gophers/graphql-go/errors"
+	"github.com/graph-gophers/graphql-go/introspection"
+	"github.com/graph-gophers/graphql-go/trace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+const (
+	graphqlFieldTag = "graphql.field"
+	graphqlQueryTag = "graphql.query"
+	graphqlTypeTag  = "graphql.type"
+)
+
+// A Tracer implements the graphql-go/trace.Tracer interface by sending traces
+// to the Datadog tracer.
+type Tracer struct {
+	cfg *config
+}
+
+// TraceQuery traces a GraphQL query.
+func (t *Tracer) TraceQuery(ctx context.Context, queryString string, operationName string, variables map[string]interface{}, varTypes map[string]*introspection.Type) (context.Context, trace.TraceQueryFinishFunc) {
+	span, ctx := tracer.StartSpanFromContext(ctx, "graphql.request",
+		tracer.ServiceName(t.cfg.serviceName),
+		tracer.Tag(graphqlQueryTag, queryString),
+	)
+
+	return ctx, func(errs []*errors.QueryError) {
+		var err error
+		if len(errs) == 1 {
+			err = errs[0]
+		} else if len(errs) > 1 {
+			msg := errs[0].Error() +
+				fmt.Sprintf(" (and %d more errors)", len(errs)-1)
+			err = errors.Errorf("%s", msg)
+		}
+		span.Finish(tracer.WithError(err))
+	}
+}
+
+// TraceField traces a GraphQL field access.
+func (t *Tracer) TraceField(ctx context.Context, label string, typeName string, fieldName string, trivial bool, args map[string]interface{}) (context.Context, trace.TraceFieldFinishFunc) {
+	span, ctx := tracer.StartSpanFromContext(ctx, "graphql.field",
+		tracer.ServiceName(t.cfg.serviceName),
+		tracer.Tag(graphqlFieldTag, fieldName),
+		tracer.Tag(graphqlTypeTag, typeName),
+	)
+	return ctx, func(err *errors.QueryError) {
+		// this is necessary otherwise the span gets marked as an error
+		if err != nil {
+			span.Finish(tracer.WithError(err))
```

Comment:
Yeah it's weird: https://play.golang.org/p/8JiXCGldIxi

```go
type e struct{}

func (*e) Error() string { return "this works" }

func main() {
	var err error = (*e)(nil)
	fmt.Println(err, err == nil, err.Error())
}
```

It happens because an interface `nil` is different from a pointer `nil`. In this example calling the `Error` method actually still works though, and I guess that's why it's allowed. 

I don't know if there's a better way to handle this upstream. We could do it with reflection, but that's probably not worth the computational overhead.

I can update the comment to make it more clear why its necessary if that would be helpful.

Reply:
I see. This could even cause us problems if people have the same expectations as me. It could result in unexpected errors where there are in fact none. There should be an issue on https://github.com/golang/go about this if there is not. I understand how this behavior is correct but can be misleading for errors. Possibly something for Go 2.

Reply:
Indeed its confusing. There's a FAQ entry and a proposal https://github.com/golang/go/issues/22729

Reply:
Based on the [FAQ](https://golang.org/doc/faq#nil_error) this looks like a beginner mistake, so I'm not sure if there is anything to do here at all, other than perhaps not make people think that it's ok to pass "_nil_" errors to `tracer.WithError`.